File: src/edu/neu/musicId/wav/WaveData.java

package edu.neu.musicId.wav;

import java.util.Collection;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * This class represents the data stored in a WAVE format file. It contains the
 * header information, as well as the collection of {@link Chunk}s that are
 * stored in the WAVE.
 * 
 * @author ckolek
 * @since 1.0
 */
public class WaveData {
    public static final String CHUNK_ID__RIFF = "RIFF";
    public static final String CHUNK_ID__RIFX = "RIFX";

    private final String chunkId;
    private final boolean isLittleEndian;
    private final long chunkSize;
    private final String format;

    private final Map<String, Chunk> chunks;

    public WaveData(String chunkId, boolean isLittleEndian, long chunkSize, String format,
            Collection<Chunk> chunks) {
        this.chunkId = chunkId;
        this.isLittleEndian = isLittleEndian;
        this.chunkSize = chunkSize;
        this.format = format;

        this.chunks = new LinkedHashMap<String, Chunk>();

        for (Chunk chunk : chunks) {
            this.chunks.put(chunk.getId(), chunk);
        }
    }

    /**
     * @return the chunk ID of the WAVE master chunk
     * @since 1.0
     */
    public String getChunkId() {
        return chunkId;
    }

    /**
     * @return <code>true</code> if the data in the WAVE is stored in
     *         little-endian form, or <code>false</code> if it is stored in
     *         big-endian form
     * @since 1.0
     */
    public boolean isLittleEndian() {
        return isLittleEndian;
    }

    /**
     * @return the chunk size (in <code>byte</code>s) of the WAVE master chunk
     * @since 1.0
     */
    public long getChunkSize() {
        return chunkSize;
    }

    /**
     * @return the format <code>String</code> of the WAVE master chunk
     * @since 1.0
     */
    public String getFormat() {
        return format;
    }

    /**
     * @param id the chunk ID of the {@link Chunk} to retrieve
     * @return the {@code Chunk} with the given chunk ID
     * @since 1.0
     */
    public Chunk getChunk(String id) {
        return chunks.get(id);
    }

    @Override
    public String toString() {
        StringBuilder string = new StringBuilder();
        string.append("chunkId: ").append(chunkId).append("\n");
        string.append("chunkSize: ").append(chunkSize).append("\n");
        string.append("format: ").append(format).append("\n");
        string.append("chunks: ").append(chunks.keySet());

        return string.toString();
    }

    /**
     * This class represents a chunk of WAVE data.
     * 
     * @author ckolek
     * @since 1.0
     */
    public static class Chunk {
        private final String id;
        private final byte[] data;

        public Chunk(String id, byte[] data) {
            this.id = id;
            this.data = data;
        }

        /**
         * @return the chunk ID of this {@code Chunk}
         * @since 1.0
         */
        public String getId() {
            return id;
        }

        /**
         * @return the length (in <code>byte</code>s) of the data in this
         *         {@code Chunk}
         * @since 1.0
         */
        public int length() {
            return data.length;
        }

        /**
         * @return the data in this {@code Chunk}
         * @since 1.0
         */
        public byte[] getData() {
            return data;
        }
    }
}

File: wave/waveData.py
__author__ = 'ckolek'

from wave.waveDataFormat import WaveDataFormat

class WaveData:
    """
    Represents all of the data found in a WAVE file including header information
     such as:
        - Chunk ID
        - Chunk Size
        - Format

     as well as all of the chunks of the data.
    """

    def __init__(self, chunk_id, is_little_endian, chunk_size, format, chunks):
        """
        :param chunk_id: the Chunk ID (4-byte string)
        :param is_little_endian: True if chunk_id == "RIFF", False if
            chunk_id == "RIFX"
        :param chunk_size: the Chunk Size (in bytes) of the WAVE data
        :param format: the WAVE format (4-byte string)
        :param chunks: a collection of the Chunks of the WAVE data
        """

        self.chunk_id = chunk_id
        self.is_little_endian = is_little_endian
        self.chunk_size = chunk_size
        self.format = format

        self.chunks = dict()

        for chunk in chunks:
            self.chunks[chunk.id] = chunk

        self._wave_data_format = None

    def chunk(self, chunk_id):
        """
        :param chunk_id: The Chunk ID of the subchunk to retrieve
        :return: the subchunk of the WAVE data with the given Chunk ID
        """

        return self.chunks[chunk_id]

    def get_wave_data_format(self):
        """
        :return: the WaveDataFormat object for this WaveData object
        """
        if self._wave_data_format is None:
            self._wave_data_format = WaveDataFormat.from_wave_data(self)

        return self._wave_data_format
    wave_data_format = property(get_wave_data_format)

    class Chunk:
        """
        Represents a subchunk of WAVE data including the Chunk ID and raw data
        """

        def __init__(self, id, data):
            """
            :param id: the Chunk ID of the subchunk
            :param data: the raw data of the subchunk
            """

            self.id = id
            self.data = data

        def length(self):
            """
            :return: the length (in bytes) of the raw data of the subchunk
            """

            return len(self.data)